backend/logic/shop/create.js

```diff
@@ -0,0 +1,351 @@
+  }
+  // Remove any leading/trailing space.
+  name = name.trim()
+
+  // Store name must only contain alpha-numeric characters and space.
+  // TODO: Add support for UTF_8 characters. This requires changes throughout the stack
+  //       and in particular encoding/decoding shop name in URLs.
+  if (!name.match(/^[a-zA-Z0-9\-' ]+$/)) {
+    return {
+      error:
+        'The shop name contains invalid character. Only alphabetical characters, numbers, space, hyphen and apostrophe are allowed.'
+    }
+  }
+  if (listingId && !String(listingId).match(/^[0-9]+-[0-9]+-[0-9]+$/)) {
+    return {
+      error: 'Listing ID must be of form xxx-xxx-xxx eg 1-001-123'
+    }
+  }
+  if (listingId && !listingId.startsWith(networkId)) {
+    return {
+      error: `Listing ID ${listingId} is not on expected Network ID ${networkId}`
+    }
+  }
+  if (!authToken) {
+    return { error: 'Provide an auth token' }
+  }
+  if (!sellerId) {
+    return { error: 'Provide a seller ID' }
+  }
+
+  const shop = await Shop.create({
+    name,
+    networkId,
+    listingId,
+    authToken,
+    config,
+    sellerId,
+    hostname
+  })
+
+  return { shop }
+}
+
+/**
+ * Creates a new shop.
+ * @param {models.Seller} seller DB object
+ * @param {string} name: Name of the store.
+ * @param {string} dataDir: Seed data dir. This is usually set to the shop's name.
+ * @param {string} sellerEmail: merchant's email.
+ * @param {string} supportEmail: Optional. Shop's support email.
+ * @param {string} web3Pk: Optional. Web3 private key for recording orders as offers on the marketplace.
+ *   Only necessary when the system operates in on-chain mode.
+ * @param {string} listingId: Optional. Marketplace listingId.
+ *   Only necessary when the system operates in on-chain mode.
+ * @param {string} printfulApi: Optional. Printful API key.
+ * @param {string} shopType: Optional. The type of shop.
+ *  The following types are supported:
+ *    'empty': Default shop type. Used the empty template.
+ *    'blank': No template used. Only creates populates the DB and not the data dir.
+ *    'local-dir': Creates a new shop that uses the data dir from an already existing shop. For super-admin use only.
+ *  The following types are not currently enabled (the BE supports them but the FE does not use them):
+ *    'clone-ipfs': Creates a new shop and clones products data from a shop IPFS hash.
+ *    'affiliate': Creates the shop using the affiliate template.
+ *    'single-product': Creates the shop using the single-product template.
+ *    'multi-product': Creates the shop using the multi-product template.
+ *    'printful': Create a new shop and populate the products by syncing from printful.
+ * @returns {Promise<{success: false, reason: string, field:string, message: string}|{success: true, slug: string}>}
+ */
+async function createShop({
+  seller,
+  name,
+  dataDir,
+  supportEmail,
+  web3Pk,
+  listingId,
+  printfulApi,
+  shopType
+}) {
```

Comment:
No objection to anything going on here, but for clarity, I might recommend breaking this down even further into more component functions.  Especially the config mergeing/building going on below.

Not a deal breaker but I think it would help with clarity.

Reply:
Yes agreed, this function is quite long. Let me land this PR and I'll keep working on better modularity.